### DISCUSSION TITLE
Using Source Generator to create simpler method overload

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#10516] [Q&A]
This question is not directly related to Avalonia per se, but someone here might have the answer. I got a class that currently needs to be registered like this (to work reflection-free)

```c#
ViewLocator = new StrongViewLocator()
    .Register<MainWindowViewModel>(new ViewDefinition(typeof(MainWindow), () => new MainWindow()))
    .Register<AddTextDialogViewModel>(new ViewDefinition(typeof(AddTextDialog), () => new AddTextDialog()));
```

I want to get the equivalent of the above code generated using this simpler syntax.

```c#
ViewLocator = new StrongViewLocator()
    .Register<MainWindowViewModel, MainWindow>()
    .Register<AddTextDialogViewModel, AddTextDialog>();
```

How can I achieve this with Source Generator?

When I think of what a generated CS file would look like, I'm really not sure. On the other hand, [Splat.Dl.SourceGenerator](https://github.com/reactiveui/Splat.DI.SourceGenerator) does that kind of synthetic sugar and more; tried looking at the source but it's very complex, without having an idea of the overall picture.

What am I getting myself into?

**Comment by wieslawsoltes** [comment]
This isn't exactly what you want but it works without reflection 
https://github.com/zkSNACKs/WalletWasabi/blob/1b8d142d0ec6892f13f6c50c3fec5d7ceb7f06ce/WalletWasabi.Fluent.Generators/StaticViewLocatorGenerator.cs#L12

Usage:
https://github.com/zkSNACKs/WalletWasabi/blob/1b8d142d0ec6892f13f6c50c3fec5d7ceb7f06ce/WalletWasabi.Fluent/ViewLocator.cs#L7

**Reply by mysteryx93** [reply]
ah, you avoid the need to manually register all the types altogether?

I like your idea.

Then I wouldn't have to choose "auto-Reflection mode" vs "strongly-typed mode", but could combine both together and not need reflection at all.

The only think I would need to add to your locator is to allow customizing the naming convention, and select View or Window based on desktop or mobile mode.

**Reply by mysteryx93** [reply]
I have a few questions about your code.

1. How is the scope of ViewModel analysis defined? Does it search only in the current assembly, or also in shared libraries?

2. You have these defined as constants in the generator. Can I have those defined in the ViewLocator instead?

    private const string ViewModelSuffix = "ViewModel";
    private const string ViewSuffix = "View";

3. By default, I select the view names according to this expression. Can I have the source generator use that from the ViewLocator?

```c#
protected virtual string GetViewName(object viewModel)
{
    const string ViewModel = "ViewModel";
    var result = viewModel.GetType().FullName!.Replace(".ViewModels.", ".Views.");
    if (result.EndsWith(ViewModel))
    {
        result = result.Remove(result.Length - ViewModel.Length) + (UseSinglePageNavigation ? "View" : "Window");
    }
    return result;
}
```

Or even better, if you could expand your class to be more generic, and work cross-platform, I'd be very grateful :)

I think that your source generator can be very useful and is worth making a NuGet package for.

**Comment by mysteryx93** [comment]
Someone pointed out that what I was asking can be done with simple extension method. How did I miss that?

```
public static StrongViewLocator Register<TViewModel, TWindow>(this StrongViewLocator viewLocator)
    where TWindow : new()
{
    return viewLocator.Register<TViewModel>(new ViewDefinition(typeof(TWindow), () => new TWindow()))
}
```

Still, I really like your option too.

Ya right, I cannot apply the "keep public constructors" attribute to TWindow, but `where TWindow : new()` does exactly that.
